Terrain / imagery rendering overhaul

kring opened the pull request:
This is _not_ ready to merge yet, but I'm opening it now to track remaining problems and ideas, some of which will go on the roadmap rather than being implemented prior to merging this branch.

Required:
* [x] I've modifed the upsample worker to start immediately rather than the first time it's needed, because in this branch it's needed much later and it's a long wait to start it up once we do realize it's needed. Not sure if this is ok. Also I've done that in such a way that it throws an exception.
* [x] Fill mesh / vertex array should be freed once the tile is loaded.
* [x] Fill tile generation allocates more memory than necessary. The worst offender is that it creates two copies of the vertices for... reasons.
* [x] Test failures.
* [x] There's debug code around (sometimes commented out).
* [x] What should the default setting of `QuadtreePrimitive.preloadAncestors` be?
* [x] ~~Test with the new BVH data from the modified server and make sure that it's still working as expected and providing a benefit.~~ Tested it and found it provided no benefit, removed it.
* [x] Test with terrain exaggeration, with/without water mask, and with all the combinations of possible vertex data, i.e. with/without normals, Web Mercator texture coordinate.
* [x] Code duplication between real loading process and fill generation (e.g. around imagery layers)
* [x] Entire globe flickers when adding a new imagery layer.

Broken Sandcastle examples:
* [x] 3D Tiles feature picking crashes trying to use destroyed vertex array after moving around a bit. Caused by draw command referencing fill tile vertex array but real vertex array has already been loaded and fill freed.
* [x] Clamp to terrain initially loads a black scene.

Nice to have:
* [x] The fill tile generation is buggy in some cases, which can lead to cracks between adjacent fill tiles and (rarely) even flickering as a fill tile switches its geometry back and forth between frames. Mostly it looks ok though.
* [x] For very coarse tiles (e.g. levels 0-4), fill tiles could use some more vertices in order to match the globe curvature. There's a fairly obvious artifact from this when spinning the globe quickly from a global view.
* [x] We're a little too aggressive in waiting for tiles to be ready before rendering them, which can lead to slow imagery layers blocking all refinement.

Future:
* [ ] We could reduce the need for fill tiles by upsampling culled tiles (but not loading their terrain or imagery) so they're ready to render when the camera moves and they're no longer culled. We'd only want to do this when the parent is loaded, though, which it really only will be (eventually) if `QuadtreePrimitive.preloadAncestors` is true.
* [ ] Currently fill tiles match adjacent vertices at their edges and have just one vertex in the midle. We could improve their quality by also copying vertices that fall inside the tile from the nearest ancestor, but then we'd need a much more complicated tessellation, like an incremental Delaunay. Using the existing triangles when all three ancestor vertices are fall inside the tile will help. Probably not worth the CPU time, but it would be interesting to investigate.

cesium-concierge commented:
Thanks for the pull request @kring!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/fill-tiles/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

pjcozzi commented on Source/Core/TerrainMesh.js:
```diff
@@ -104,7 +127,252 @@ define([
+            var v = transformTextureCoordinate(clipRectangleVMin, clipRectangleVMax, uv.y);
+            var uOrV = compareU ? u : v;
+
+            var inside = uOrV >= 0.0 && uOrV <= 1.0;
```
Just style, but here and throughout, please wrap the right hand side of these expressions in parentheses.

pjcozzi commented on Source/Core/TerrainTileEdgeDetails.js:
```diff
@@ -0,0 +1,17 @@
+define([], function() {
+    'use strict';
+
+    function TerrainTileEdgeDetails() {
```
Just style here and throughout, but I think we explicitly mark these as `@private` enough though it isn't required.

E.g. https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Cesium3DTilesetStatistics.js

pjcozzi commented on Source/Scene/GlobeSurfaceTile.js:
```diff
@@ -120,6 +142,28 @@ define([
+    GlobeSurfaceTile.prototype.getBvh = function(tile, terrainProvider) {
+        if (this._bvh === undefined) {
+            var terrainData = this.terrainData;
+            if (terrainData !== undefined && terrainData.bvh !== undefined) {
```
Throughout, should these be using `!defined()` for consistency even if it is old school?

kring replied:
> Throughout, should these be using !defined() for consistency even if it is old school?

I'm torn. I mean, checking for null when it will only ever be undefined is a waste of time, right? Not like this is public API stuff. But yeah, we usually use `defined()` and I didn't have a really good reason to do differently here.

pjcozzi commented on Source/Scene/GlobeSurfaceTile.js:
```diff
@@ -239,55 +265,57 @@ define([
+                    // The ancestor that holds the BVH data isn't loaded yet; load it (terrain only!) instead of this tile.
+                    GlobeSurfaceTile.processStateMachine(ancestor, frameState, terrainProvider, imageryLayerCollection, vertexArraysToDestroy, true);
+                    return;
+
```
Extra whitespace...because I clearly have no important review comments.

pjcozzi commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -84,16 +100,45 @@ define([
+     * are not yet loaded / renderable.
+     * @private
+     */
+    var MissingTileStrategy = {
```
Will these options just be replaced with the one right/passable strategy?

kring replied:
Yeah I'll probably get rid of this. There are some other tweakables I'll leave in, though.

pjcozzi commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -508,8 +579,13 @@ define([
         GlobeSurfaceTile.processStateMachine(tile, frameState, this._terrainProvider, this._imageryLayers, this._vertexArraysToDestroy);
         var tileLoadedEvent = this._tileLoadedEvent;
+
+        // TODO: creating a new function for every loaded tile every frame?!
```
Please please please do not.  😢 

I went to great pains to avoid this in the early 3D Tiles engine.

kring replied:
This creation of a new function every frame is already in master. I was surprised to see it and left this TODO to remind myself to maybe do something about it.

pjcozzi commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -128,12 +133,42 @@ define([
+         * @type {Number}
+         * @default 20
+         */
+        this.loadingDescendantLimit = 20;
```
Does this number need more tweaking?  Or are you convinced that 20 is close enough?

kring replied:
I just added a Terrain Tweaks sandcastle example (under develoment) if you feel like playing with it. 20 feels about right to me, but there's no right answer, it's just balancing user feedback against performance.

pjcozzi replied:
Cool, I will try to make time but no promises.

pjcozzi commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -128,12 +133,42 @@ define([
+
+        /**
+         * Gets or sets a value indicating whether the siblings of rendered tiles should be preloaded.
+         * Setting this to true
```
I guess you know this doc is a TODO.

pjcozzi commented on Source/Scene/TerrainFillMesh.js:
```diff
@@ -0,0 +1,1273 @@
+        var startIndex, endIndex, existingTile, existingRectangle;
+        var sourceRectangle = sourceTile.rectangle;
+
+        function firstIsLessThanSecond(a, b, epsilon) {
```
Do these beyond in `CesiumMath` with friends?

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Math.js#L569

pjcozzi commented on terrainnotes.md:
```diff
@@ -0,0 +1,113 @@
+To refine past a tile, one of the following conditions must be met:
```
Would be great to move these to https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation

Maybe in a new `ImplementationNotes` or whatever subdirectory.

kring replied:
Mostly just a brain dump and probably not worth keeping around. But maybe I'll turn the most useful parts into a blog post or something.

pjcozzi replied:
Blog post would be great!

kring commented:
Not ready for a final review yet though, due to the things in the checklist above.

pjcozzi commented:
@kring please bump when ready.  @tfili and @bagnell will take a look.

cesium-concierge commented:
Thanks again for your contribution @kring! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

mramato commented:
@OmarShehata we should probably tweak the concierge to look at last commit, not last comment.

OmarShehata commented:
@mramato yup! That's next on my list (https://github.com/AnalyticalGraphicsInc/cesium-concierge/issues/60).

jbo023 commented:
Hi @kring 
I just did a first test of your terrain/imagery branch with our datasets. 
I really like the results. On the startingscreen of our demo application https://berlin.virtualcitymap.de/#/ 
the requested terrain tiles went down from 105  to 46.
The requested Imagery tiles from 113 to 66.

Unfortunately the branch does not work with ie11. 
The problem is that the internet explorer does not support slice operations on UInt16Arrays.

https://github.com/AnalyticalGraphicsInc/cesium/blob/fill-tiles/Source/Workers/createVerticesFromQuantizedTerrainMesh.js#L137

kring commented:
Thanks for trying the branch @jbo023! Glad to hear you're seeing good results with it. The IE11 crash is fixed now, thanks for reporting it.

kring commented:
This is getting really close. I was just looking at one of the last TODO items:

> Test with the new BVH data from the modified server and make sure that it's still working as expected and providing a benefit.

And to my surprise, the BVH data from the server doesn't help at all! It doesn't reduce the number of tiles or speed up loading. I spent a long time trying to understand why, and it seems to be that the metrics for selecting tiles even without an accurate BVH are just so good now 😎.

A typical scenario is:
1. Based on min/max heights inherited from an ancestor, it looks like we want tile level 16 to meet our SSE.
2. But loading level 16 tiles based on that would be a bad idea, because we'd probably load a lot of useless tiles.
3. We could back off to level 15 where we have BVH data and load that, but it could be risky, too. If our min/max heights came from level 0, for example, we could end up loading a lot of useless level 15 tiles, too.
4. So what we do instead is estimate distance based on the farthest-away height (min or max) of the tiles we don't have accurate min/max heights for. Real tiles are guaranteed to be closer than this. In our example, this results in selection of a level 13 tiles.
5. Now we still don't know 100% that our level 13 tile is visible. But we do know that if it's not culled then it or it descendants will meet our SSE. We have two options:
  a. load the level 13 tile (terrain only) and find out. This is what we do when BVH data isn't available at all, or
  b. load the nearest BVH node (level 10 in this example) and that will tell us whether the level 13 tile is really visible.

But with (b), we still have to load that level 13 tile if the level 10 tile tells us it's visible, so why do the extra load? We might decide to load deeper tiles intead based on the BVH in that level 10 tile, but loading the level 13 tile would probably accomplish that as well by shrinking the min/max height range for the deeper tiles.

The only real benefit of (b) is that the BVH information it contains is applicable to _other_ level 13 tiles (and also some level 11, 12, 14, and 15 tiles). But even that isn't as useful as it sounds because that level 13 tile usually covers a fairly large area because of how we underestimated its SSE, so the levels 11,12, and 13 tiles aren't really relevant and the levels 14 and 15 descendants of our 13 are already suitably improved by loading that 13.

Maybe panning mid-load could see some benefit. I wasn't able to measure one, though.

So I'm planning to rip out all the BVH stuff from the client unless someone can think of a reason why I shouldn't.

pjcozzi commented:
> I'm planning to rip out all the BVH stuff from the client unless someone can think of a reason why I shouldn't.

@tfili I assume this is separate from any existing code - if any - using the BVH to workaround a large tileset.json?

pjcozzi commented:
@tfili ah, gotcha!

kring commented:
There's a remaining problem in the Clamp to Terrain Sandcastle example, and I think I might need some help to figure it out. The initial "Draw Point with Label" works fine. Switch to "Draw Billboard" and it works well, too. Switch back to "Draw Point with Label" and Cesium zooms in _way_ too close, so close that the label's glyph rendering doesn't work right and it looks like this:
![image](https://user-images.githubusercontent.com/924374/51725539-ddde9780-20b6-11e9-9801-6a0ca0e09c12.png)

After spending all day trying to figure this out, here's what I know:
* `adjustBoundingSphereOffset` in `Camera.js` sets the camera to be 100 meters (`MINIMUM_ZOOM`) when the bounding sphere's radius is exactly 0.0. That's what we get the first time around and in master where this works.
* In the problematic situation above, the radius is ~0.05 meters, so instead Cesium uses the similarly named but very different `camera._scene.screenSpaceCameraController.minimumZoomDistance`.
* The radius is small but not zero because the bounding sphere is the combination of the label's bounding sphere and the point's bounding sphere. Both of these bounding spheres have a radius of 0.0 but their centers are 0.1 meters away from each other.
* [here's where things get murky]
* It seems that an existing label is re-used from an `EntityCluster`. But then the clamping isn't updated because the position (unclamped) is not different from the last time this label was used.
* I don't know, and haven't been able to figure out, why the clamped position would be different given the same unclamped position and same terrain.
* A lot of the code in `Globe.getHeight` is duplicated in the `updateHeights` function in `QuadtreePrimitive`, so a difference between them that results in ever-so-slightly-different answers for the clampled position would account for this, but I haven't been able to pinpoint it.
* There's a lot of back and forth between cartesian and cartographic coordinates in the clamping/height updating code, so an error introduced somewhere in that process would account for this, but I haven't been able to pinpoint it.

I can "fix" this bug by changing `BoundingSphere.fromBoundingSpheres` to treat the nearly-zero-radius union of two zero-radius spheres as having a zero radius itself. But that's dodgy.

Anyone have any ideas? @bagnell especially since I think you wrote a lot of the clamping code?

kring commented:
Well, I pushed a fix for the above that I'm a little happier with, but it still seems kinda dodgy. Mind taking a good look at https://github.com/AnalyticalGraphicsInc/cesium/pull/7061/commits/e250a39b71f42877a9e8c09533204391dba0cd05 @bagnell?

pjcozzi commented:
@bagnell or perhaps @hpinkos or @lilleyse the above would be a good one to look at soon to keep momentum on this PR. https://github.com/AnalyticalGraphicsInc/cesium/pull/7061#issuecomment-457475865

kring commented on Source/Scene/TerrainFillMesh.js:
```diff
@@ -0,0 +1,1176 @@
+            nextIndex = addEdge(fill, ellipsoid, encoding, typedArray, nextIndex, fill.northTiles, fill.northMeshes, TileEdge.SOUTH);
+
+            // Add a single vertex at the center of the tile.
+            // TODO: minimumHeight and maximumHeight only reflect the corners
```
Address TODO

kring commented on Source/Scene/TerrainFillMesh.js:
```diff
@@ -0,0 +1,1176 @@
+            Cartesian3.normalize(normal, normal);
+            AttributeCompression.octEncode(normal, vertex.encodedNormal);
+        } else {
+            // TODO: do we need to actually compute a normal?
```
Address TODO

kring commented on Source/Workers/upsampleQuantizedTerrainMesh.js:
```diff
@@ -532,5 +531,7 @@ define([
 
-    return createTaskProcessorWorker(upsampleQuantizedTerrainMesh);
+    var worker = createTaskProcessorWorker(upsampleQuantizedTerrainMesh);
+    worker._workerFunction = upsampleQuantizedTerrainMesh;
```
This isn't used anymore, remove it.

kring commented on terrainnotes.md:
```diff
@@ -0,0 +1,113 @@
+To refine past a tile, one of the following conditions must be met:
```
Remove / clean up / move these notes.

kring commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -619,6 +668,145 @@ define([
+        var cameraPosition = frameState.camera.positionWC;
+        var cameraDirection = frameState.camera.directionWC;
+        var tileDirection = Cartesian3.normalize(Cartesian3.subtract(obb.center, cameraPosition, tileDirectionScratch), tileDirectionScratch);
+        return (1.0 - Cartesian3.dot(tileDirection, cameraDirection)) * tile._distance;
```
It's not a huge difference, but I believe this metric is subjectively better (more detail loads in better places sooner) than a simple distance metric.

kring commented:
Performance measured using `Terrain Performance.html`, modified to use the built/minified version of Cesium. Before each measurement I reloaded the page and cleared my browser's cache before pressing the test button.

# Static Horizontal

Measurement | master | fill-tiles
----------------|---------|-------------
Time (seconds) | 10.53 | 8.82
Requests | 824 | 578
Min FPS | 17.6 | 24.0
Max FPS | 30.7 | 34.8
Avg FPS | 24.6 | 28.2

kring commented:
I think this is finally ready! :tada:
I'll record and attach some side-by-side loading GIFs later tonight. I'm very happy with the reduced cracking during load in this branch, too.

NaderCHASER commented:
This branch is live at https://maps.allisonhouse.com/free/maps.html and provides a considerable improvement to our use case.

Thanks for all your hard work @kring

OmarShehata commented on Apps/Sandcastle/gallery/development/Terrain Performance.html:
```diff
@@ -0,0 +1,195 @@
+//     requestWaterMask : true,
+//     requestVertexNormals : true
+// });
+// viewer.terrainProvider = cesiumTerrainProviderMeshes;
```
Is this commented code still needed here anymore? Was this just for testing out the modified terrain/bvh?

OmarShehata commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -130,9 +142,16 @@ define([
+         * highlighted at all. The alpha value is used to alpha blend with the tile's
+         * actual color.
+         */
+        this.fillHighlightColor = undefined;
```
This is just for debugging right? Should this be named something like `debugFillHightlightColor` to be consistent with similar things like [Cesium3DTileset.debugColorizeTiles](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Cesium3DTileset.js#L669) ?

kring replied:
No, I don't see this as only for debugging. I actually considered having it on by default, and I think it might be a good idea to enable it for certain kinds of applications. Fill tiles don't represent the real terrain surface at all, they just fill the space. So it can be useful to visually indicate that a part of the surface is complete nonsense. For example, you might want to darken fill tiles a bit, or maybe even make them appear 100% black. The latter is almost like not rendering the files at all, except that you can't see the stars and atmosphere artifacts through black fill tiles. The effect is similar to what you get in 2D maps when you pan quickly and adjacent tiles aren't loaded yet.

OmarShehata commented on Source/Renderer/Texture.js:
```diff
@@ -267,6 +267,15 @@ define([
+     */
+    Texture.create = function(options) {
+        return new Texture(options);
+    };
```
This seems a bit unusual in that it changes how you have to create a new Texture to be able to mock it. I couldn't other tests that try to mock a constructor, but I think the more standard thing would be to do a `Texture._implementations.create` which can be mocked, and the constructor can just essentially call that. That way the calling site can still be `new Texture`. 

Resource.js does this https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Resource.js#L1758.

OmarShehata commented on Specs/TerrainTileProcessor.js:
```diff
@@ -0,0 +1,123 @@
+    'use strict';
+
+    function TerrainTileProcessor(frameState, terrainProvider, imageryLayerCollection) {
+        this.frameState = frameState;
```
This is a mock class for testing right? How come it's not called `MockX` in the same way the others like `MockTerrainProvider` and `MockImageryProvider` are named?

OmarShehata commented on Specs/createGlobe.js:
```diff
@@ -25,7 +25,6 @@ define([
             },
             _surface : {},
-            tileLoadedEvent : new Event(),
             imageryLayersUpdatedEvent : new Event(),
```
I couldn't really figure this out - was `tileLoadedEvent` just removed? If so, why?

OmarShehata replied:
Can you elaborate a bit more on _"was extremely costly"_ ? Like just raising this event is costly or does it involve having to keep track of extra things? I think it's probably fine if there is such a high cost to it.

kring replied:
It was mostly this code:
https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/GlobeSurfaceTileProvider.js#L518

That was creating a new function for every loaded tile every frame.

I didn't spent a lot of time trying to think of an efficient way to implement this because it didn't seem very useful to me.

OmarShehata commented on Source/Scene/Globe.js:
```diff
@@ -650,7 +645,14 @@ define([
+
+        // For low-detail tiles, large triangles often cut the the globe and appear to be at a much
+        // lower height than actually makes any sense. So clamp the height to the actual height range
+        // of the tile.
```
Is this the issue where you can see stark elevation differences between neighbouring tiles while the terrain is loading in? Is there an easy way to recreate this or see this clamping height in action?

kring replied:
Originally, terrain clamping didn't work at all in this branch. I added this thinking it would help, but it didn't, and I had to add the `_neededPositions` system instead. So we could probably remove it. 

It's a pretty reasonable change, though. If a terrain tile is known to contain heights beween -100 and 8700 meters, and you ask for a height at a point within that tile, it seems reasonable that it wouldn't return something less than -100. Without this check, level 0 tiles will commonly return some very negative numbers.

On the other hand, _without_ this change, the returned height will actually be on the tile's triangles. With this change, it will sometimes be floating above them (even if imperceptibly because of the SSE-driven selection).

So ¯\_(ツ)_/¯ . Anyone have a strong opinion?

OmarShehata commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -489,210 +562,393 @@ define([
+    function TraversalDetails() {
+        this.allAreRenderable = true;
+        this.anyWereRenderedLastFrame = false;
+        this.notYetRenderableCount = 0;
```
I can see that these seem like a significant part of the traversal strategy. Do you think it deserves some doc comment here? It's unclear for example what it keeps counting `notYetRenderableCount` for.

OmarShehata commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -109,6 +117,12 @@ define([
+        // even though it is not rendered. The first position will be the position of the
+        // camera and the second will be the origin of the camera's reference frame.
+        this._neededPositions = [undefined, undefined];
+
```
Why is `_neededPositions` an array instead of an object with `_neededPositions.camera` and `_neededPositions.cameraReferenceOrigin` ? I think as an array it's more extendable and I assume it can be used for more positions in the future. It just seems potentially error-prone to be making this assumption about what the first and second positions will always be.

kring replied:
Yeah, exactly, the idea is that other positions can be added and get the same treatment. But the first two are "reserved" for a specific purpose. Could be over engineering.

OmarShehata commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -754,17 +1009,28 @@ define([
-        processSinglePriorityLoadQueue(primitive, frameState, tileProvider, endTime, tileLoadQueueMedium);
-        processSinglePriorityLoadQueue(primitive, frameState, tileProvider, endTime, tileLoadQueueLow);
+        var didSomething = processSinglePriorityLoadQueue(primitive, frameState, tileProvider, endTime, tileLoadQueueHigh, false);
+        didSomething = processSinglePriorityLoadQueue(primitive, frameState, tileProvider, endTime, tileLoadQueueMedium, didSomething);
```
`didSomething` is a bit vague. Maybe it should be something like `loadedTile` ?

kring replied:
Changed it to `didSomeLoading`. The idea is to make sure that at least one tile gets a chance to make some load progress. Previously one tile per priority group was given a chance to load, which was less than ideal.

OmarShehata commented on Source/Scene/TileImagery.js:
```diff
@@ -43,13 +43,14 @@ define([
      * @returns {Boolean} True if this instance is done loading; otherwise, false.
      */
-    TileImagery.prototype.processStateMachine = function(tile, frameState) {
+    TileImagery.prototype.processStateMachine = function(tile, frameState, skipLoading) {
```
Does `skipLoading` really mean `skipLODs` ? Or is it something else? If it's the former, should it be renamed?

kring replied:
No, nothing to do with skipping LODs. I'll change the doc to hopefully make it clearer.

OmarShehata commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -128,12 +142,44 @@ define([
+         * @default true
+         */
+        this.preloadAncestors = true;
+
```
This and `preloadSiblings` seem like pretty useful and easy to understand parameters, but there's currently no way to modify these from the public API right? (I know the development/Terrain Tweaks Sandcastle shows toggling them). I think these might be worth exposing, maybe on `Globe.js`, and adding them as checkboxes to the `Terrain.html` Sandcastle.

kring replied:
I added these properties to `Globe`. Added `loadingDescendantLimit` and `fillHighlightColor` too. I'm a little hesitant to add them to `Terrain.html` though, cause I think they'll just make terrain look overcomplicated there.

kring commented:
Thanks for the review @OmarShehata, I think I've addressed all your comments.

> Just to summarize my understanding of what's going on here, this PR implements both skipping LODs for terrain/imagery as well as being able to extrapolate neighbor terrain tile geometry until the terrain tiles load (that's what "fill tiles" means), and this is great because then we don't have to load sibling tiles, because the fill can make up for that.

Yep, exactly. Really it's the fill tiles that make the level skipping possible. I mean, we could render a single scene with a fixed camera with skipping and not need any fill tiles, and that would work fine.

The problem comes in when the camera moves. If 3 out of 4 tiles in a quad are visible, and we load those, we can render the scene. But then if the camera moves so that 4th is visible, what do we do? We can't render a parent or other ancestor, because those three tiles we used to be rendering would suddenly get less detailed and it would look terrible. Especially if the only loaded tile we have to fill that space is, say, a level 0 tile.

We can upsample that 4th tile, but upsampling is pretty expensive. We'd have to do it before the camera moves, rather than sychronously after it moves. Plus upsampling, say, a level 15 tiles from a level 0 tile gives some pretty terrible results (huge cracks).

So the fill tiles are a quick, not-too-terrible-looking thing we can create synchronously just in time when we realize the camera has moved and we suddenly need to draw something in that space. And have that at our disposal makes level skipping possible.

>  it doesn't seem like fill tiles works in the water

These weren't gaps, just much darker shading because the water effect was missing. It's fixed now.

OmarShehata commented:
Thanks for the quick response @kring ! The water tiles work well now, and I'm happy with the new tweaks. One thing you forgot is to add a type and a default to `preloadSiblings` and `fillHighlightColor` to the doc on `Globe.js`. I always run `npm run generateDocumentation` to make sure it looks right when deployed.

@tfili and/or @bagnell would you like to review this now? To make things easier, check out this [comment](https://github.com/AnalyticalGraphicsInc/cesium/pull/7061#issuecomment-460217748) for a nice summary of what all is going on here.

The below is also a list of links to comments that are outstanding issues/open questions:

* [Should neededPositions be an array?](https://github.com/AnalyticalGraphicsInc/cesium/pull/7061#discussion_r253205571)
  * I think we should just convert it to an object now, and if we need to do this for more positions in the future, we can change it without breaking anything since it's all private.
* [Should we clamp the returned heights?](https://github.com/AnalyticalGraphicsInc/cesium/pull/7061#discussion_r253336419)
  * I would say no, assuming that most people use these heights to clamp something to terrain visually, and that would make it correspond to the actual rendered geometry even if it's outside the range of the terrain tile's reported heights.
* [Deprecating tileLoadedEvent](https://github.com/AnalyticalGraphicsInc/cesium/pull/7061#discussion_r253335918)
* [Spying on Texture's constructor](https://github.com/AnalyticalGraphicsInc/cesium/pull/7061#discussion_r253335606)
  * It's unusual but it's a convincing argument.

kring commented:
Ok I simplified `_neededPositions` (changed it to two fields on `QuadtreePrimitive`) and removed the height clamping from `getHeight`. That should be everything up to now.

kring commented:
Here's the loading GIF I promised:
![fill-tiles-static-horizontal2](https://user-images.githubusercontent.com/924374/52324989-78d05d80-2a37-11e9-8848-21c6303cea42.gif)

loshjawrence commented:
Looks great @kring, saw similar request and timing savings on my end. Looking at some master vs branch chrome profilers sessions, the only thing that stuck out was an uptick on "getProgramParemeter" during shader compilation, they are rare but also large. Nothing that should hold back this PR but something to bring to your attention in case you may have any ideas about it. 

![getprogramparameter](https://user-images.githubusercontent.com/13598291/52374334-fe014580-2a2a-11e9-8a7e-29bd773a224b.png)

One other thing, though minor, was the QuadTreePrimitive.render where some traversal is done, the overall total time for all those function calls went up:

![somediff](https://user-images.githubusercontent.com/13598291/52374601-9b5c7980-2a2b-11e9-937b-c2afcd48f68b.png)

tfili commented:
I was zooming around in the San Diego coastal area where we have our most detailed terrain. I was exploring the coast for a minute or two and this popped up

```
DeveloperError: This object was destroyed, i.e., destroy() was called.
Error
    at new DeveloperError (http://localhost:8080/Source/Core/DeveloperError.js:43:19)
    at Texture.throwOnDestroyed (http://localhost:8080/Source/Core/destroyObject.js:45:19)
    at createFillMesh (http://localhost:8080/Source/Scene/TerrainFillMesh.js:720:28)
    at TerrainFillMesh.update (http://localhost:8080/Source/Scene/TerrainFillMesh.js:75:13)
    at addDrawCommandsForTile (http://localhost:8080/Source/Scene/GlobeSurfaceTileProvider.js:1541:30)
    at GlobeSurfaceTileProvider.endUpdate (http://localhost:8080/Source/Scene/GlobeSurfaceTileProvider.js:479:17)
    at QuadtreePrimitive.render (http://localhost:8080/Source/Scene/QuadtreePrimitive.js:390:26)
    at Globe.render (http://localhost:8080/Source/Scene/Globe.js:791:21)
    at updateAndRenderPrimitives (http://localhost:8080/Source/Scene/Scene.js:2945:26)
    at executeCommandsInViewport (http://localhost:8080/Source/Scene/Scene.js:2778:13)
```

kring commented:
I wasn't able to reproduce that @tfili, but looking at the code I can _sort of_ see how that might happen, especially if you were switching between terrain sources with and without water. I made a code change that should account for that, let me know if you see it again.

kring commented:
Thanks for profiling @loshjawrence. I can't think of any reason why `getProgramParameter` would take longer in this branch, but I'll think about it. I take it you're seeing that consistently in this branch, and not at all in `master`? 🤔 

Tile selection taking longer isn't too surprising, though. For one thing, deep refinement happens earlier in this branch, which will mean more tiles to visit. And it also does more computation to select the tiles to load and render (i.e. the "kick" mechanism). Am I right in thinking it's still a _reasonably_ small part of the overall rendering time, though?

pjcozzi commented:
> Thanks for profiling @loshjawrence. I can't think of any reason why `getProgramParameter` would take longer in this branch, but I'll think about it. I take it you're seeing that consistently in this branch, and not at all in `master`? 🤔

Shader compile/link can be async so it is possible that `getProgramParameter` is being called right after compile/link WebGL calls and then blocking until the compile/link happens - so the real issue could be that the shader cache is trashing or whatever and shaders are being compiled more often.

loshjawrence commented:
Yes to both questions

tfili commented:
In the Google Earth sandcastle example I can reproduce a crash everytime. I just zoom in and try and tilt to the horizon and I get this
```
TypeError: Cannot read property 'length' of undefined
TypeError: Cannot read property 'length' of undefined
    at getCornerFromEdge (http://localhost:8080/Source/Scene/TerrainFillMesh.js:1161:30)
    at getCorner (http://localhost:8080/Source/Scene/TerrainFillMesh.js:864:13)
    at createFillMesh (http://localhost:8080/Source/Scene/TerrainFillMesh.js:495:24)
    at TerrainFillMesh.update (http://localhost:8080/Source/Scene/TerrainFillMesh.js:75:13)
    at addDrawCommandsForTile (http://localhost:8080/Source/Scene/GlobeSurfaceTileProvider.js:1541:30)
    at GlobeSurfaceTileProvider.endUpdate (http://localhost:8080/Source/Scene/GlobeSurfaceTileProvider.js:479:17)
    at QuadtreePrimitive.render (http://localhost:8080/Source/Scene/QuadtreePrimitive.js:390:26)
    at Globe.render (http://localhost:8080/Source/Scene/Globe.js:791:21)
    at updateAndRenderPrimitives (http://localhost:8080/Source/Scene/Scene.js:2945:26)
    at executeCommandsInViewport (http://localhost:8080/Source/Scene/Scene.js:2778:13)
```

tfili commented:
@kring I think this all looks good to me besides the one sandcastle crash. I went through all the sandcastle examples that use terrain. I made sure things played nice with the new metadata availability. Seems to work well in IE and Firefox as well.

kring commented:
Thanks @tfili, I fixed the Google Earth crash. I _think_ the terrain GEE terrain is working correctly, but it's very low res. Maybe take a quick look and make sure it looks ok to you, too.

kring commented:
> so the real issue could be that the shader cache is trashing or whatever and shaders are being compiled more often.

I can't see any evidence of extra shader compilation/linking happening in this branch.

@loshjawrence can you tell me more about how you acquired that flame chart? Which app? What were you doing at the time? Anything else relevant?

I did some profiling myself, and I occasionally see some stupid-long frame times (one was 200ms!), but I can't understand how to interpret the flame chart. For example, here's an 82.7ms frame where the majority of the time was spent doing.... absolutely nothing? What do the huge gaps mean?

![image](https://user-images.githubusercontent.com/924374/52475885-1cb13900-2bf1-11e9-9d0a-d9fdb4804123.png)

In any case, I haven't seen `getProgramParameter` take a long time, so I must not be going what you're doing.

jbo023 commented:
We are using this branch in our development environment and didn't find any issues yet

tfili commented:
Thanks @kring. The seems to work. I tested with a higher res, non-public server and everything looks great.

mramato commented:
Hey @kring!  Awesome work here. I was hoping to use a small app I wrote to test tiling/streaming performance to check out this branch and I ran into a bug.  It looks like `scene.globe.tilesLoaded` doesn't get set to true when all tiles are loaded anymore (at least in some cases).

Basically the app loads a few scenes, dumps stats to the console and then puts up an alert when it's finished.  To run it:

1. Unzip this app: [CesiumPerformance.zip](https://github.com/AnalyticalGraphicsInc/cesium/files/2845714/CesiumPerformance.zip)
2. `npm install`
3. `npm start`
4. Navigate to http://127.0.0.1:9090

The app works fine when pointing at the latest release (which it currently does), but if you update package.json to point to the CI package from your branch (http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/fill-tiles/cesium-1.54.0-fill-tiles30451.tgz) the tests never finish because test number 2 never gets to `tileLoaded=true`.

It's using combined but unminified Cesium, so it's hopefully not too hard to debug in-app if the problem is not obvious.  I didn't try updating it to use the models, but that should be easy enough for you to do if you need it.

loshjawrence commented:
The flame chart was via chrome devtools like the one you posted. It was on the Terrain performance.html and the button that did the moving forward across the terrain. From what I remember the gaps just mean chrome devtools isn't instrumented for whatever is going at that time.

kring commented:
@mramato thanks for picking up on that, it's fixed now. The terrain engine now loads the tile that contains the camera's position and the tile that contains the camera's reference frame's origin (if it's not the center of the Earth) even if those tiles would otherwise be culled. This is necessary to make terrain clamping and camera collision detection work reliably when skipping levels.

But when loading these "culled but needed" tiles, we only load the terrain, not the imagery. As a result, these tiles were being put in the load queue every frame because they weren't done loading, but they never got done loading because we purposely avoided loading imagery. The fix is to only add them to the load queue if the _terrain_ isn't done loading, rather than adding them if they need any loading at all.

mramato commented:
Awesome, @kring, thanks.  Do you have planned changes left on your end, or are you just waiting for more feedback?

Does anyone else have any feedback here?  I'm going to take another look now, but barring any issues I think we want to get this merged today or tomorrow to give it as much time in master as possible.  Thanks!

mramato commented:
@kring I can confirm your update fixed the problem I was seeing with loading, thanks again.

Now I'm seeing some unexpected behavior when setting `RequestScheduler.throttleRequests = false` 

In 1.54, if throttleRequests is false, `RequestScheduler.statistics.numberOfActiveRequestsEver` is always equal to `RequestScheduler.statistics.numberOfAttemptedRequests`, since no requests are ever cancelled.  In this branch, `numberOfAttemptedRequests` is still slightly larger (though not as large as when throttling is enabled).

As expected, `numberOfCancelledRequests` remains at 0 both in master and this branch, so I believe throttling is indeed disabled.  Perhaps this is just an accounting issue somewhere?

mramato commented:
Just noticed cancelled requests is 0 even with throttling on, so something changed under the hood that maybe I'm just not aware of?

mramato commented:
Here are some rough numbers comparing 1.54 and this branch across several different views.  Bandwidth/data savings are just as impressive as runtime loading performance.  Once this is merged into master, I'll probably re-run these tests again with some additional vetting just to make sure everything is on the up-and-up.

* Numbers are from "best of 3" runs.
* Canvas size of 1920x1080
* Used combined but (not minified) Cesium (next time I'll use minified)
* Default CesiumWidget with Sentinel-2 Imagery and Cesium World Terrain (no water/lighting) from ion.
* Request throttling enabled (the default)
* All tiles from the local cache (to remove network overhead).
* Tests ran in Chrome 72.0.3626.96 (Official Build) (64-bit) on Ubuntu 18.04.
* Times are from call to `camera.setView` until `scene.globe.tilesLoaded` gets set to true.

Also note, since I used locally cached tiles, this is the "best" case for master.  With a clear cache I would expect master would fair even worse since it has to request more tiles.

Overall, this branch kicks ass, but the savings can be vastly different from view to view. With the views I used, this branch seems to load the tests 1.4x to 2.3x times faster and uses ~33% less data/bandwidth overall.  There might be some better views to test with as well.

From the intangibles, this branch also loads things in a way that makes it appear even faster to end users, no more back to front waiting for the world to pop-in.  We can't discount user impression when talking about the hard numbers.

My testing methodology or assumptions might get some tweaks, since I have some open questions to Kevin about RequestStatistics differences in this branch.

#### 1.54

Overall: 19.7 MB of tiles loaded. (Via Chrome Network tab)

Grand canyon horizon: 4818 ms
Number of server requests:  352
Number of attempted requests: 1386
Number of cancelled requests: 755
Percentage of throttled requests: 74.60

Everest: 5034 ms
Number of server requests:  544
Number of attempted requests: 4269
Number of cancelled requests: 1734
Percentage of throttled requests: 87.25

AGI top-down: 2787 ms
Number of server requests:  238
Number of attempted requests: 770
Number of cancelled requests: 211
Percentage of throttled requests: 69.09


#### fill-tiles

Overall: 13.3 MB of tiles loaded. (Via Chrome Network tab)

Grand canyon horizon: 3391 ms
Number of server requests:  210
Number of attempted requests: 635
Number of cancelled requests: 0
Percentage of throttled requests: 66.92

Everest: 3850 ms
Number of server requests:  369
Number of attempted requests: 635
Number of cancelled requests: 0
Percentage of throttled requests: 41.88

AGI top-down: 1166ms
Number of server requests:  98
Number of attempted requests: 132
Number of cancelled requests: 0
Percentage of throttled requests: 25.75

mramato commented:
Played with a bunch of Sandcastle examples and didn't run into any other problems, so this PR gets a :+1: from me (outside of that one question about request statistics).

hpinkos commented on Source/Core/Math.js:
```diff
@@ -583,6 +583,108 @@ define([
+     *          <code>absoluteEpsilon<code>. <code>false</code> if <code>left</code> is greater or if the two
+     *          values are nearly equal.
+     */
+    CesiumMath.leftIsLessThanRight = function(left, right, absoluteEpsilon) {
```
Should these just be `lessThan`, `lessThanOrEqualTo`, etc?  We don't call the function `leftIsEqualToRight`

kring replied:
I thought may names made the user clearer. Order doesn't matter in `equals`, so being precise is less important there. But if you all feel strongly about it I'm happy to rename them.

OmarShehata replied:
I do like this because otherwise I'd have to go look up the documentation whenever using this to check which way it's comparing them. This way it's self-documenting.

hpinkos replied:
Right than left is standard, so this seems a little pedantic.  In `Check` we just call these functions `greaterThan` `lessThan` etc https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Check.js#L89

OmarShehata replied:
That's fair. Sounds like we should change it for consistency here then @kring .

tfili replied:
I honestly don't care either way, but `Check` isn't public so I wouldn't necessarily use that as the decision maker.

mramato replied:
I thought we had functions like this all over the code, but to might surprise, JulianDate is the only place I could find them (outside of Check).  While I'm on board with renaming them for consistency, I don't want it to hold up this PR which really needs time to stew in master.

If someone wants to open a separate PR to clean things up before the next release, I'll review and merge it ASAP.

kring commented:
@mramato yeah `RequestScheduler` isn't being used to throttle requests any more in this branch. I mean, it does still drop requests where there are already too many in flight to a particular hostname, but it no longer does any canceling or reordering for terrain or imagery. I removed it because:

* The mechanism was really expensive. There was a seprate function, with its own closure, created per tile 😬. Plus all the calling distance computation functions and sorting every frame.
* It produced bad results. `RequestScheduler` only knew how to sort by distance. The terrain engine uses an improved prioritization system that slightly favors tiles near the center of the screen. It also has low/medium/high priority queues. Ignoring the queues and sorting by distance causes us to load a bunch of really unimportant low priority tiles earlier than the important ones, just because they happen to be closer.

More discussion of this from back in June:
https://github.com/AnalyticalGraphicsInc/cesium/pull/6243#issuecomment-401224681

kring commented:
Oh and yeah I think everything has been addressed up to this point.

mramato commented:
> @mramato yeah `RequestScheduler` isn't being used to throttle requests any more in this branch.

Fine by me, I actually never liked what we had anyway.  @lilleyse @loshjawrence @OmarShehata were you already aware of this and are you okay with it? Any concerns that would prevent us from merging this?

Assuming no one else has anything, I'm fine with merging this today.

loshjawrence commented:
The last time I set throttle to false it totally trashed streaming performance for camera-moving scenarios on 3dtiles. I guess I can copy the Request and RequestScheduler files over and remeasure some things.

mramato commented:
> The last time I set throttle to false it totally trashed streaming performance for camera-moving scenarios on 3dtiles. I guess I can copy the Request and RequestScheduler files over and remeasure some things.

Thanks.  Just let us know ASAP if there are any actionable concerns for this PR going into master, while I'm sure it might affect some of the branches you have in-flight, unless there's some sort of architectural showstopper, that shouldn't hold up this PR.

loshjawrence commented:
Using the changes from Request and RequestScheduler in this PR on our staging branch looks harmless from my testing. Any diff's seem to be within the noise.

Orig:
attempted: 32962 cancelled: 2793 flight loads: 277 site time: 15.28
attempted: 26767 cancelled: 3600 cancelled active: 35   flight loads: 473 site time: 3.54
attempted: 24293 cancelled: 4422 cancelled active: 138 flight loads: 505 site time: 5.11
Total Time: 23.94

Ring's changes:
attempted: 33636 cancelled: 2881 flight loads: 277 site time: 15.28 
attempted: 26695 cancelled: 3760 cancelled active: 32 flight loads: 461 site time: 3.66
attempted: 24370 cancelled: 4573 cancelled active: 136 flight loads: 515 site time: 5.25
Total Time: 24.20

OmarShehata commented:
@mramato I think this gets a ✔️ from me as well. From the earlier discussion @kring pointed to about the request scheduler it sounds inline with the direction we're going.

lilleyse commented:
@mramato Good with me as well.

mramato commented:
Since this branch was a bit out of date, I merged in master.  Assuming it passes without any surprises, I'll merge this up.

mramato commented:
@kring are there any new GitHub issues we should write up (such as those unchecked boxes?)

Other than that, I'm going to merge.  I encourage everyone to continue to put things through their paces in master and also report any unusual problems you might find over the next few weeks.

mramato commented:
While writing up a forum post, I was amused by the fact that this overhaul is going to be released exactly 6 years after initial terrain support was added (3/1/13 in version b14): https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md#b14---2013-03-01

Some guy named @kring did all the work on it back then.  How times have changed.
